assembly/src/tokens/lines.rs

```diff
@@ -0,0 +1,323 @@
+fn token_lines_works_multiline_tokens() {
+    let source = r#"begin add # foo
+            mul end"#;
+    let tokens: Vec<_> = TokenLines::from(source).collect();
+    let expected = [
+        SourceToken::new(SourceLocation::new(1, 1), "begin"),
+        SourceToken::new(SourceLocation::new(1, 7), "add"),
+        SourceToken::new(SourceLocation::new(2, 13), "mul"),
+        SourceToken::new(SourceLocation::new(2, 17), "end"),
+    ]
+    .to_vec();
+    assert_eq!(tokens, expected);
+}
+
+#[test]
+fn token_lines_works_simple_mod_comment() {
+    let source = r#"#! some mod comment      
+    begin add # foo
+
+            # bar
+
+            mul
+      end
+
+
+            # baz"#;
+    let tokens: Vec<_> = TokenLines::from(source).collect();
+    let expected = [
+        SourceToken::new(SourceLocation::new(2, 5), "begin")
+            .with_docs(["some mod comment"].to_vec()),
+        SourceToken::new(SourceLocation::new(2, 11), "add"),
+        SourceToken::new(SourceLocation::new(6, 13), "mul"),
+        SourceToken::new(SourceLocation::new(7, 7), "end"),
+    ]
+    .to_vec();
+    assert_eq!(tokens, expected);
+}
+
+#[test]
+fn token_lines_works_void_remainder_doc_comment() {
+    let source = r#"#! some mod comment      
+    begin add # foo
+
+            # bar
+
+            mul
+      end
+
+
+            #! baz"#;
+    let tokens: Vec<_> = TokenLines::from(source).collect();
+    let expected = [
+        SourceToken::new(SourceLocation::new(2, 5), "begin")
+            .with_docs(["some mod comment"].to_vec()),
+        SourceToken::new(SourceLocation::new(2, 11), "add"),
+        SourceToken::new(SourceLocation::new(6, 13), "mul"),
+        SourceToken::new(SourceLocation::new(7, 7), "end"),
+    ]
+    .to_vec();
+    assert_eq!(tokens, expected);
+}
+
+#[test]
+fn token_lines_works_inline_doc_comment() {
+    let source = r#"#! some mod comment      
+    begin add # foo
+
+            #! bar
+
+            mul
+    #!                    end doc comment with trailing spaces                
+      end
+
+
+            #! baz"#;
+    let tokens: Vec<_> = TokenLines::from(source).collect();
+    let expected = [
+        SourceToken::new(SourceLocation::new(2, 5), "begin")
+            .with_docs(["some mod comment"].to_vec()),
+        SourceToken::new(SourceLocation::new(2, 11), "add"),
+        SourceToken::new(SourceLocation::new(6, 13), "mul").with_docs(["bar"].to_vec()),
+        SourceToken::new(SourceLocation::new(8, 7), "end")
+            .with_docs(["end doc comment with trailing spaces"].to_vec()),
+    ]
+    .to_vec();
+    assert_eq!(tokens, expected);
+}
+
+#[test]
+fn token_lines_works_inline_multiline_doc_comment() {
+    let source = r#"#! some mod comment      
+        #!      additional docs
+    begin add # foo
+
+            #! bar
+
+            mul
+    #!                    end doc comment with trailing spaces                
+            #! more lines....
+      end
+
+
+            #! baz"#;
```

Comment:
This test should fail - doc comments must either be at the beginning of a module or be always be followed by a procedure declaration.